# Mercedes Benz Project

# Step 1 - Importing the required Libraries & test/train datasets

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
import pandas as pd
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')


In [7]:
df_train.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [8]:
df_train.shape

(4209, 378)

In [9]:
df_test.head()

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   
3   4  az  l   n  f  d  z  l  n    0  ...     0     0     0     1     0     0   
4   5   w  s  as  c  d  y  i  m    0  ...     1     0     0     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  
3     0     0     0     0  
4     0     0     0     0  

[5 rows x 377 columns]

In [10]:
df_test.shape

(4209, 377)

In [11]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 378 entries, ID to X385
dtypes: float64(1), int64(369), object(8)
memory usage: 12.1+ MB


In [12]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 377 entries, ID to X385
dtypes: int64(369), object(8)
memory usage: 12.1+ MB


In [13]:
df_train.describe()

ID            y          X10     X11          X12  \
count  4209.000000  4209.000000  4209.000000  4209.0  4209.000000   
mean   4205.960798   100.669318     0.013305     0.0     0.075077   
std    2437.608688    12.679381     0.114590     0.0     0.263547   
min       0.000000    72.110000     0.000000     0.0     0.000000   
25%    2095.000000    90.820000     0.000000     0.0     0.000000   
50%    4220.000000    99.150000     0.000000     0.0     0.000000   
75%    6314.000000   109.010000     0.000000     0.0     0.000000   
max    8417.000000   265.320000     1.000000     0.0     1.000000   

               X13          X14          X15          X16          X17  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.057971     0.428130     0.000475     0.002613     0.007603  ...   
std       0.233716     0.494867     0.021796     0.051061     0.086872  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     1.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.318841     0.057258     0.314802     0.020670     0.009503   
std       0.466082     0.232363     0.464492     0.142294     0.097033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.007603     0.001663     0.000475     0.001426  
std       0.089524     0.086872     0.040752     0.021796     0.037734  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 370 columns]

In [14]:
df_test.describe()

ID          X10          X11          X12          X13  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean   4211.039202     0.019007     0.000238     0.074364     0.061060   
std    2423.078926     0.136565     0.015414     0.262394     0.239468   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%    2115.000000     0.000000     0.000000     0.000000     0.000000   
50%    4202.000000     0.000000     0.000000     0.000000     0.000000   
75%    6310.000000     0.000000     0.000000     0.000000     0.000000   
max    8416.000000     1.000000     1.000000     1.000000     1.000000   

               X14          X15          X16          X17          X18  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.427893     0.000713     0.002613     0.008791     0.010216  ...   
std       0.494832     0.026691     0.051061     0.093357     0.100570  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       1.000000     0.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.325968     0.049656     0.311951     0.019244     0.011879   
std       0.468791     0.217258     0.463345     0.137399     0.108356   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.008791     0.000475     0.000713     0.001663  
std       0.089524     0.093357     0.021796     0.026691     0.040752  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 369 columns]

# Drop the ID column from dataset

In [15]:
df_train.drop("ID",inplace= True, axis=1)
df_test.drop("ID",inplace= True, axis=1)

In [16]:
df_train.describe()

y          X10     X11          X12          X13  \
count  4209.000000  4209.000000  4209.0  4209.000000  4209.000000   
mean    100.669318     0.013305     0.0     0.075077     0.057971   
std      12.679381     0.114590     0.0     0.263547     0.233716   
min      72.110000     0.000000     0.0     0.000000     0.000000   
25%      90.820000     0.000000     0.0     0.000000     0.000000   
50%      99.150000     0.000000     0.0     0.000000     0.000000   
75%     109.010000     0.000000     0.0     0.000000     0.000000   
max     265.320000     1.000000     0.0     1.000000     1.000000   

               X14          X15          X16          X17          X18  ...  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  ...   
mean      0.428130     0.000475     0.002613     0.007603     0.007840  ...   
std       0.494867     0.021796     0.051061     0.086872     0.088208  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       1.000000     0.000000     0.000000     0.000000     0.000000  ...   
max       1.000000     1.000000     1.000000     1.000000     1.000000  ...   

              X375         X376         X377         X378         X379  \
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000   
mean      0.318841     0.057258     0.314802     0.020670     0.009503   
std       0.466082     0.232363     0.464492     0.142294     0.097033   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

              X380         X382         X383         X384         X385  
count  4209.000000  4209.000000  4209.000000  4209.000000  4209.000000  
mean      0.008078     0.007603     0.001663     0.000475     0.001426  
std       0.089524     0.086872     0.040752     0.021796     0.037734  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 369 columns]

In [17]:
df_test.head()

X0 X1  X2 X3 X4 X5 X6 X8  X10  X11  ...  X375  X376  X377  X378  X379  \
0  az  v   n  f  d  t  a  w    0    0  ...     0     0     0     1     0   
1   t  b  ai  a  d  b  g  y    0    0  ...     0     0     1     0     0   
2  az  v  as  f  d  a  j  j    0    0  ...     0     0     0     1     0   
3  az  l   n  f  d  z  l  n    0    0  ...     0     0     0     1     0   
4   w  s  as  c  d  y  i  m    0    0  ...     1     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     0     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 376 columns]

# If for any column(s), the variance is equal to zero, then you need to remove those variable(s).

In [18]:
zero_variance_columns_train = df_train.var()[df_train.var() == 0].index.values


In [19]:
zero_variance_columns_train

array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

In [20]:
zero_variance_columns_test = df_test.var()[df_test.var() == 0].index.values

In [21]:
zero_variance_columns_test

array(['X257', 'X258', 'X295', 'X296', 'X369'], dtype=object)

In [22]:
df_train.drop(zero_variance_columns_train, inplace=True, axis=1)
df_test.drop(zero_variance_columns_test, inplace=True, axis=1)

#  Conclusion= Columns having variance zero is dropped

In [23]:
df_train.shape

(4209, 365)

In [24]:
df_test.shape

(4209, 371)

# Check for null and unique values for test and train sets.

In [25]:
np.sum(df_train.isnull().sum())


0

In [26]:
np.sum(df_test.isnull().sum())

0

In [27]:
df_train.isnull().sum()

y       0
X0      0
X1      0
X2      0
X3      0
       ..
X380    0
X382    0
X383    0
X384    0
X385    0
Length: 365, dtype: int64

# Conclusion= No Null values in test & train datasets

# Which Columns are non-numeric? Unique

In [31]:
df_train.describe(include=['object'])

X0    X1    X2    X3    X4    X5    X6    X8
count   4209  4209  4209  4209  4209  4209  4209  4209
unique    47    27    44     7     4    29    12    25
top        z    aa    as     c     d     v     g     j
freq     360   833  1659  1942  4205   231  1042   277

In [32]:
np.unique(df_train.X3)


array(['a', 'b', 'c', 'd', 'e', 'f', 'g'], dtype=object)

In [33]:
New_columns = df_train.describe(include=['object']).columns.values

In [34]:
New_columns

array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

# Apply label encoder

In [35]:
from sklearn.preprocessing import LabelEncoder

In [36]:
encoder = LabelEncoder()

In [37]:
for col in New_columns:
 encoder.fit(df_train[col].append(df_test[col]).values)
 df_train[col] = encoder.transform(df_train[col])
 df_test[col] = encoder.transform(df_test[col])


In [38]:
df_train.head(5)

y  X0  X1  X2  X3  X4  X5  X6  X8  X10  ...  X375  X376  X377  X378  \
0  130.81  37  23  20   0   3  27   9  14    0  ...     0     0     1     0   
1   88.53  37  21  22   4   3  31  11  14    0  ...     1     0     0     0   
2   76.26  24  24  38   2   3  30   9  23    0  ...     0     0     0     0   
3   80.62  24  21  38   5   3  30  11   4    0  ...     0     0     0     0   
4   78.02  24  23  38   5   3  14   3  13    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 365 columns]

In [39]:
df_test.head(5)

X0  X1  X2  X3  X4  X5  X6  X8  X10  X11  ...  X375  X376  X377  X378  \
0  24  23  38   5   3  26   0  22    0    0  ...     0     0     0     1   
1  46   3   9   0   3   9   6  24    0    0  ...     0     0     1     0   
2  24  23  19   5   3   0   9   9    0    0  ...     0     0     0     1   
3  24  13  38   5   3  32  11  13    0    0  ...     0     0     0     1   
4  49  20  19   2   3  31   8  12    0    0  ...     1     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 371 columns]

# Perform dimensionality reduction. There are so many columns which can be reduced

In [46]:
from sklearn.decomposition import PCA

In [48]:
pcaModel = PCA(0.98, svd_solver='full')
X=df_train.drop('y', axis=1)
y=df_train['y']
pcaModel.fit(X)

PCA(copy=True, iterated_power='auto', n_components=0.98, random_state=None,
    svd_solver='full', tol=0.0, whiten=False)

In [49]:
pcaModel.n_components_


12

# So in 12 columns we can accomate the 98% information, original datset contains 378 columns

In [51]:
pcaModel.explained_variance_ratio_

array([0.40868988, 0.21758508, 0.13120081, 0.10783522, 0.08165248,
       0.0140934 , 0.00660951, 0.00384659, 0.00260289, 0.00214378,
       0.00209857, 0.00180388])

In [53]:
np.mean(pcaModel.explained_variance_ratio_)

0.08168017413045565

# check if exp var ratio is greater than threshold value

In [55]:
pcaModel.explained_variance_ratio_ > np.mean(pcaModel.explained_variance_ratio_) 

array([ True,  True,  True,  True, False, False, False, False, False,
       False, False, False])

In [76]:
from sklearn.model_selection import train_test_split

In [77]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.3)

In [78]:
pca = PCA(n_components=9,svd_solver='full')
X_train = pca.fit_transform(X_train)
X_test = pca.fit_transform(X_test)


In [80]:
df_test.shape

(4209, 371)

In [81]:
pca_df_test = pca.fit_transform(df_test)

In [82]:
X_train.shape, X_test.shape,pca_df_test.shape

((2946, 9), (1263, 9), (4209, 9))

# Predict your test_df values using XGBoost

In [83]:
#pip install xgboost

In [84]:
import xgboost as xgb

In [85]:
xgb_model = xgb.XGBRegressor(objective='reg:squarederror',learning_rate = 0.1)

In [86]:
xgb_model.fit(X_train,y_train)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.1, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=100, n_jobs=0, num_parallel_tree=1,
             objective='reg:squarederror', random_state=0, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, subsample=1, tree_method='exact',
             validate_parameters=1, verbosity=None)

In [87]:
pred = xgb_model.predict(X_test)

In [88]:
print('Training Accuracy:', xgb_model.score(X_train,y_train))
print('Test Accuracy:', xgb_model.score(X_test,y_test))

Training Accuracy: 0.742242555431986
Test Accuracy: -0.3529889082560826


In [89]:
from sklearn.metrics import r2_score,mean_squared_error
from math import sqrt

In [90]:
print('R2 Score:', r2_score(y_test,pred))
print('RMSE:', np.sqrt(mean_squared_error(y_test,pred)))

R2 Score: -0.35298890825608265
RMSE: 14.398647269304881


In [91]:
y_pred = xgb_model.predict(pca_df_test)

In [92]:
print('R2 Score:', r2_score(y_test,pred))
print('RMSE:', np.sqrt(mean_squared_error(y_test,pred)))

R2 Score: -0.35298890825608265
RMSE: 14.398647269304881


# Lets use and try Randomforest regressor to inprove R2 score

In [93]:
from sklearn.ensemble import RandomForestRegressor
regressor = xgb.XGBRegressor(objective='reg:linear', learning_rate=0.1)

In [123]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

In [124]:
Features_train = pd.DataFrame(pcaModel.transform(X_train))

In [125]:
Features_train.describe()

0            1            2            3            4   \
count  3367.000000  3367.000000  3367.000000  3367.000000  3367.000000   
mean     -0.026799    -0.211816     0.029763    -0.065402    -0.063075   
std      15.648904    11.364890     8.933146     8.056159     7.043068   
min     -23.071496   -22.697173   -17.883045   -19.616685   -13.583307   
25%     -13.331227    -7.660939    -8.274029    -6.497304    -5.920797   
50%      -4.380650    -2.516139     0.897065    -1.182782    -0.603521   
75%      12.427669     6.220204     7.167600     6.542448     5.957571   
max      39.025839    32.845313    18.839900    20.409804    14.308816   

                5            6            7            8            9   \
count  3367.000000  3367.000000  3367.000000  3367.000000  3367.000000   
mean      0.015586    -0.007407    -0.006182     0.000833     0.009954   
std       2.896704     1.995723     1.528100     1.261684     1.143447   
min      -5.040693    -4.377823    -3.289078    -3.196062    -2.844281   
25%      -2.130067    -1.941183    -1.503764    -0.993944    -0.634118   
50%      -0.205659     0.408907     0.405359    -0.028099    -0.053136   
75%       1.240799     1.497018     1.179973     0.808253     0.521783   
max       7.519440     4.685064     3.447685     3.612572     5.094022   

                10           11  
count  3367.000000  3367.000000  
mean      0.000478    -0.002169  
std       1.124331     1.030432  
min      -3.058761    -2.234745  
25%      -0.775005    -0.624894  
50%      -0.046273    -0.102529  
75%       0.729930     0.442617  
max       3.547067     4.431119

In [126]:
FeaturesAndTarget_train = Features_train
FeaturesAndTarget_train['y'] = y_train

In [128]:
import xgboost as xgb
from sklearn.ensemble import RandomForestRegressor
regressor = xgb.XGBRegressor(objective='reg:linear', learning_rate=0.1)

In [129]:
Features_train.columns

Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 'y'], dtype='object')

In [130]:
Features_train.drop(columns=['y'], inplace= True, axis=1)

In [148]:
Features_train.columns

Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], dtype='object')

In [149]:
regressor.fit(Features_train, y_train)

[20:17:18] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.2.0/src/objective/regression_obj.cu:174: reg:linear is now deprecated in favor of reg:squarederror.
[20:17:19] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.2.0/src/objective/regression_obj.cu:174: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.1, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=100, n_jobs=0, num_parallel_tree=1,
             objective='reg:linear', random_state=0, reg_alpha=0, reg_lambda=1,
             scale_pos_weight=1, subsample=1, tree_method='exact',
             validate_parameters=1, verbosity=None)